### PR TITLE
[12.0][FIX] bi_view_editor: window functions without ORDER BY clause do not have reliable behavior

### LINKED ISSUE #388: [12.0] bi_view_editor: row_number over () does not guarantee ordering
Hi,
given the following view created using module `bi_view_editor`:
```
CREATE OR REPLACE VIEW public.x_bve_commissioni
AS SELECT row_number() OVER ()::integer AS id,
    t1.id AS x_bve_t1_id, -- manually added for analysis
    t1.agent_id AS x_bve_t1_agent_id,
    t1.partner_id AS x_bve_t1_partner_id,
    t1.date_invoice AS x_bve_t1_date_invoice,
    t1.product_id AS x_bve_t1_product_id,
    t1.price_unit AS x_bve_t1_price_unit,
    t1.commission_id AS x_bve_t1_commission_id,
    t1.price_subtotal_signed AS x_bve_t1_price_subtotal_signed,
    t1.amount AS x_bve_t1_amount
   FROM sale_commission_analysis_report t1; -- defined in https://github.com/OCA/commission/blob/befb9470a352edc9a2946bbf1c3b5a57e02badeb/sale_commission/report/sale_commission_analysis_report.py#L7
```
I am currently encountering the following strange behavior:
```
prod_db=# select * from x_bve_commissioni where id = 5130;
  id  | x_bve_t1_id | x_bve_t1_agent_id | x_bve_t1_partner_id | x_bve_t1_date_invoice | x_bve_t1_product_id |  x_bve_t1_price_unit   | x_bve_t1_commission_id | x_bve_t1_price_subtotal_signed | x_bve_t1_amount 
------+-------------+-------------------+---------------------+-----------------------+---------------------+------------------------+------------------------+--------------------------------+-----------------
 5130 |       36303 |             14343 |               19833 | 2020-03-16            |                   2 | 0.00000000000000000000 |                     11 |                           0.00 |            0.00
(1 row)

prod_db=# select id, x_bve_t1_id,  x_bve_t1_date_invoice from x_bve_commissioni where id = 5130;
  id  | x_bve_t1_id | x_bve_t1_date_invoice 
------+-------------+-----------------------
 5130 |       19991 | 2020-01-27
(1 row)
```
I expect the column `x_bve_t1_id` to be the same in both query results.

I'm no PostgreSQL expert but I did some research and the solution I came up with is adding an `order by t1.id` clause in the `OVER` of  column `row_number() OVER ()::integer AS id`.
I've seen that the column's creation is hardcoded in https://github.com/OCA/reporting-engine/blob/124d1db26b3f13f3b4669896699a4b9f3e16eb49/bi_view_editor/models/bve_view.py#L292 so I don't really know how this could be fixed.

Any ideas?
Thanks!

### PR DESCRIPTION
**Pull request by SimoRubi**
Fix #388.
An order by clause is highly recommended when using window functions:
> Without ORDER BY, rows are processed in an unspecified order.

(from https://www.postgresql.org/docs/10/sql-expressions.html#SYNTAX-WINDOW-FUNCTIONS)

This modification allows a (technical) user to specify such condition and others, like PARTITION BY.

### CODE REVIEW

**Review by legalsylvain — APPROVED** [review]
make sense. 
Thanks.

File: bi_view_editor/models/bve_view.py
Comment (astirpe) [inline bi_view_editor/models/bve_view.py]: You can remove this line `string="Over condition",` since Odoo creates the field name automatically.

File: bi_view_editor/models/bve_view.py
Comment (astirpe) [inline bi_view_editor/models/bve_view.py]: Could you add extra information regarding how to use this functionality? Or maybe some example of use?

**Review by astirpe — CHANGES_REQUESTED** [review]
@SimoRubi Thank you for this!

File: bi_view_editor/models/bve_view.py
Comment (SimoRubi) [reply bi_view_editor/models/bve_view.py]: Thanks for the review!
Yes I know that the `string` is automatically filled but I usually prefer to keep it explicit.

File: bi_view_editor/models/bve_view.py
Comment (SimoRubi) [reply bi_view_editor/models/bve_view.py]: Added an example, generic extra information (PARTITION BY, RANGE, ...) would be way too long for an `help` attribute I think.

File: bi_view_editor/models/bve_view.py
Comment (astirpe) [reply bi_view_editor/models/bve_view.py]: As per https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#fields, the string should be removed. It's also useful to not specify any string in some cases. For example, in case someone wants to manually set a different string directly in the field definition, this won't conflict when the module is updated.

File: bi_view_editor/models/bve_view.py
Comment (astirpe) [reply bi_view_editor/models/bve_view.py]: Thanks. Any useful extra explanation could be added in the readme.

File: bi_view_editor/models/bve_view.py
Comment (SimoRubi) [reply bi_view_editor/models/bve_view.py]: Removed

**Comment by astirpe** [comment]
In case the user writes an invalid over condition, the entire query would fail. In this case can be raised any user friendly warning?

File: bi_view_editor/models/bve_view.py
Comment (SimoRubi) [reply bi_view_editor/models/bve_view.py]: Added

**Comment by SimoRubi** [comment]
This is a good idea for an improvement to the module, but I think it is beyond this PR's purpose.

**Comment by astirpe** [comment]
Ok if you please add "Raise a warning in case the user writes an invalid over condition." in the `ROADMAP.rst` then I will merge immediately.

**Comment by SimoRubi** [comment]
hah, that's blackmailing! :wink:
Added, and broadened the scope of the warning to "in case the SQL query is not correct.": let's not point fingers to the new arrived field :smile:

**Review by astirpe — APPROVED** [review]

**Comment by astirpe** [comment]
I just noticed that I cannot merge since I'm not a maintainer of the module :thinking:

File: bi_view_editor/models/bve_view.py
Comment (astirpe) [inline bi_view_editor/models/bve_view.py]: What do you think about:


```python
from psycopg2.errors import Error, UndefinedObject

        # ...

        try:
            self.env.cr.execute(
                "CREATE or REPLACE VIEW %s as (%s)", (AsIs(view_name), AsIs(query),)
            )
        except UndefinedObject:
            raise ValidationError(_("The generated query is invalid. Is the OVER condition correct?"))
        except Error as e:
            raise ValidationError(_("The generated query is invalid:\n%s") % e.pgerror)
```

Comment (SimoRubi) [reply bi_view_editor/models/bve_view.py]: Ahah still blaming my dear `over condition`? I know it is the weak spot because it is the only free text input in the query definition, but maybe it would be better to print also the original exception in both cases (maybe only the `except Error` should suffice).

Moreover, I rapidly checked your solution and `psycopg2.errors` is introduced in 2.8 (ref. https://www.psycopg.org/docs/errors.html) but minimal requirement is `2.7.3.1` (ref. https://github.com/odoo/odoo/blob/012e4a3eb11feabe5634ab806ec5dd2012ffda00/requirements.txt#L28)..

Anyway I still think these changes deserve a separate issue/PR

Comment (astirpe) [reply bi_view_editor/models/bve_view.py]:
```python
        try:
            self.env.cr.execute(
                "CREATE or REPLACE VIEW %s as (%s)", (AsIs(view_name), AsIs(query),)
            )
        except psycopg2.Error as e:
            raise ValidationError(_("The generated query is invalid:\n%s") % e.pgerror)

```

Comment (SimoRubi) [reply bi_view_editor/models/bve_view.py]: Haven't tried that out, but looks good

**Review by SimoneVagile — APPROVED** [review]

**Comment by SimoRubi** [comment]
@OCA/reporting-engine-maintainers can this be merged? 
I'd like to implement the changes suggested in https://github.com/OCA/reporting-engine/pull/391#pullrequestreview-406343248 in a separate PR, after this one is merged to avoid conflicts.

**Comment by astirpe** [comment]
@SimoRubi  the PRs approvals in OCA are usually long and debated. This PR already has 3 approvals. Would you consider to include that changes here? It would save a lot of time to all of us.

**Comment by SimoRubi** [comment]
> @SimoRubi the PRs approvals in OCA are usually long and debated. This PR already has 3 approvals. Would you consider to include that changes here? It would save a lot of time to all of us.

The approvals are for the changes that are in the PR right now; if I added new changes, I should also re-request the reviews.

I have created https://github.com/OCA/reporting-engine/pull/397 for the new changes you requested

**Comment by astirpe** [comment]
What I proposed was my review for this PR. Now we have two PRs for the same issue, and this one has 3 approvals, the other has 0 ...

**Comment by SimoRubi** [comment]
That is the point, it's two separate issues:

1 (this PR) allows the user to specify something in the `OVER` condition
2 (the other PR) gives a better error message when the query is incorrect

Now, this PR as you said is ready for merge because the feature it adds has already been reviewed.

The other one might be independent from this one but the fastest way that I've found to break the query is using `over_condition` field, so that PR actually depends on this one.

**Review by astirpe — CHANGES_REQUESTED** [review]
I'm reviewing this PR.

Just tested in Runbot.

Steps to reproduce:

1. Enter an invalid `over_condition`.

Result:


```
Error:
Odoo Server Error

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/http.py", line 656, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/odoo/http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/.repo_requirements/odoo/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/.repo_requirements/odoo/odoo/http.py", line 698, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/odoo/http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/service/model.py", line 98, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 519, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/OCB-12.0/addons/web/controllers/main.py", line 966, in call_button
    action = self._call_kw(model, method, args, {})
  File "/home/odoo/OCB-12.0/addons/web/controllers/main.py", line 954, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 759, in call_kw
    return _call_kw_multi(method, model, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 746, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/odoo/build/OCA/reporting-engine/bi_view_editor/models/bve_view.py", line 381, in action_create
    self._create_sql_view()
  File "/home/odoo/build/OCA/reporting-engine/bi_view_editor/models/bve_view.py", line 293, in _create_sql_view
    AsIs(view_name), AsIs(query), ))
  File "/.repo_requirements/odoo/odoo/sql_db.py", line 148, in wrapper
    return f(self, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/sql_db.py", line 225, in execute
    res = self._obj.execute(query, params)
psycopg2.ProgrammingError: window "test" does not exist
LINE 1: ...EW x_bve_test as (SELECT   CAST(row_number() OVER (test) as ...
                                                             ^


```

**Comment by SimoRubi** [comment]
Improving the error message is a general improvement for the module, that's why I added
```
* Raise a warning in case the SQL query is not correct.
```
in the roadmap.
Anyway that does look like the correct behavior: you write an invalid query and you get an error, the added changes only give a 'better' error.

In the end, since here we are getting into a loop I'll merge the changes from https://github.com/OCA/reporting-engine/pull/397 and request again all the reviews

**Review by astirpe — CHANGES_REQUESTED** [review]
Thanks!

`over_condition` should be readonly when not in draft, otherwise it's possible to modify it while the bi view is already created. Just tested in Runbot.

**Review by astirpe — APPROVED** [review]
@legalsylvain would you merge this PR?

**Comment by SimoRubi** [comment]
@OCA/reporting-engine-maintainers can this be merged? Thanks!

**Comment by SimoRubi** [comment]
@OCA/reporting-engine-maintainers can this be merged?
2 revisions should be more than enough since this is a Beta module (ref. https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/oca_module_lifecycle_development_status.rst#2alpha-and-beta-modules).

**Comment by SimoRubi** [comment]
@OCA/reporting-engine-maintainers can someone please take a look?
@legalsylvain there is a little change (https://github.com/OCA/reporting-engine/pull/397) from when you reviewed, can you update your review?

Thanks!

**Pull request merged**
